### PR TITLE
Add buffering to `Substream` using `circular-buffer`

### PR DESCRIPTION
**Pull request by Vurich**

### CODE REVIEW

File: multiplex-rs/src/lib.rs [hunk 1]
```diff
@@ -335,21 +339,35 @@ impl<T: AsyncRead + AsyncWrite> StreamMuxer for Multiplex<T> {
     }
 }
 
+pub type MultiplexConfig = BufferedMultiplexConfig<[u8; 0]>;
```
Comment (tomaka) [inline multiplex-rs/src/lib.rs]: I find it annoying that you can no longer build a config by just writing `let cfg = MultiplexConfig;`. Instead you have to do `MultiplexConfig::default()`. But that's not a big problem.

File: multiplex-rs/src/read.rs [hunk 2]
```diff
@@ -76,14 +77,89 @@ fn create_buffer(capacity: usize) -> bytes::BytesMut {
...
+    O: Into<Option<(u32, &'a mut [u8])>>,
+    T: AsyncRead,
+>(
+    lock: &mut ::shared::MultiplexShared<T, Buf>,
     stream_data: O,
+) -> io::Result<usize> {
+    read_stream_internal(lock, stream_data.into())
+}
+
+pub fn read_stream_internal<T: AsyncRead, Buf: Array<Item = u8>>(
```
Comment (tomaka) [inline multiplex-rs/src/read.rs]: This shouldn't be `pub` I guess.

Comment (Vurich) [reply multiplex-rs/src/read.rs]: Good catch

**Comment by Vurich** [comment]
@tomaka I'm putting tests in and then I'll cover it in comments. I think if you can't follow the code then that means that I probably need more comments

**Pull request merged**